docs/site/migration/express-middleware.md

```diff
@@ -6,8 +6,82 @@ sidebar: lb4_sidebar
 permalink: /doc/en/lb4/migration-express-middleware.html
 ---
 
-{% include note.html content="
-This is a placeholder page, the task of adding content is tracked by the
-following GitHub issue:
-[loopback-next#3947](https://github.com/strongloop/loopback-next/issues/3947)
-" %}
+Migrating app-level Express middleware from LoopBack 3 (LB3) to LoopBack 4 (LB4)
+requires a wrapper Express app, which will the mount the LB4 and also act as the
+point for mounting the shared Express middleware.
+
+First, make sure you have done the necessary steps to mount your LB3 app on a
+LB4 app as described in
+[this tutorial](https://github.com/strongloop/loopback-next/tree/master/examples/lb3-application#tutorial).
+
+Then, create a directory which will be the wrapper Express app. Move the LB4 app
+into this directory, and move the `package.json` file from the LB4 app to this
+directory.
```

Comment:
I find this proposal problematic for several reasons.

1. It's not consistent with the solution we are describing in [Creating an Express Application with LoopBack REST API](https://loopback.io/doc/en/lb4/express-with-lb4-rest-tutorial.html#create-a-facade-express-application). This inconsistency may create confusion among our users, and also make support more time consuming for us - we will have to learn which Express-mounting approach is the developer using.

2. I find the changes too involved. Mounting LB4 app in a top-level Express application is a short-term workaround until we implement first-class support (see https://github.com/strongloop/loopback-next/issues/1293). Once that feature is landed, we will want the user to migrate away from top-level Express app and use LB4 as the top-level container. Ideally, such change should be mostly about removing code that's no longer needed, it should not require moving pretty much all LB files to a different location!

3. Finally, `lb4` CLI expects LoopBack files to be in the project root, alongside with `package.json`. This is important e.g. when adding a new datasource, where we want to add the connector to `package.json` dependencies.

Personally, I'd like the migration guide to follow the project layout described in our existing docs, where the LoopBack application stays in the project root directory and Express-specific things are implemented in newly added source files.

Reply:
Based on the issue description, the layout should resemble:
```
express app
 - Express middleware
 - lb4-app
   - lb3-app
```
But from this comment:
> Personally, I'd like the migration guide to follow the project layout described in our existing docs, where the LoopBack application stays in the project root directory and Express-specific things are implemented in newly added source files.

It seems as though it should be
```
lb4-app
 - Express middleware
 - lb3-app
```

Can you clarify if this is what you mean?

But I do agree the `package.json` should remain at the lb4 app level.

Reply:
Sorry for the confusion. The layout description is describing composition of Express/LB3/LB4 blocks at runtime.

The layout of project files should follow the usual LB4 project conventions, as described in _ Creating an Express Application with LoopBack REST API_.